src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedProducer.cs

```diff
@@ -0,0 +1,183 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Diagnostics;
+using System.Text;
+using Confluent.Kafka;
+using OpenTelemetry.Context.Propagation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Instrumentation.ConfluentKafka;
+
+internal sealed class InstrumentedProducer<TKey, TValue> : IProducer<TKey, TValue>
+{
+    private readonly IProducer<TKey, TValue> producerImplementation;
+    private readonly TextMapPropagator propagator = Propagators.DefaultTextMapPropagator;
+    private readonly ConfluentKafkaProducerInstrumentationOptions<TKey, TValue> options;
+
+    public InstrumentedProducer(IProducer<TKey, TValue> producer, ConfluentKafkaProducerInstrumentationOptions<TKey, TValue> options)
+    {
+        this.producerImplementation = producer;
+        this.options = options;
+    }
+
+    public Handle Handle => this.producerImplementation.Handle;
+
+    public string Name => this.producerImplementation.Name;
+
+    public int AddBrokers(string brokers)
+    {
+        return this.producerImplementation.AddBrokers(brokers);
+    }
+
+    public void SetSaslCredentials(string username, string password)
+    {
+        this.producerImplementation.SetSaslCredentials(username, password);
+    }
+
+    public async Task<DeliveryResult<TKey, TValue>> ProduceAsync(
+        string topic,
+        Message<TKey, TValue> message,
+        CancellationToken cancellationToken = default)
+    {
+        if (this.options.Traces)
+        {
+            using Activity? activity = this.StartActivity("publish", topic) ?? Activity.Current;
+            this.InjectActivity(activity, message);
+        }
+
+        return await this.producerImplementation.ProduceAsync(topic, message, cancellationToken).ConfigureAwait(false);
+    }
+
+    public async Task<DeliveryResult<TKey, TValue>> ProduceAsync(
+        TopicPartition topicPartition,
+        Message<TKey, TValue> message,
+        CancellationToken cancellationToken = default)
+    {
+        if (this.options.Traces)
+        {
+            using Activity? activity = this.StartActivity("publish", topicPartition) ?? Activity.Current;
+            this.InjectActivity(activity, message);
+        }
+
+        return await this.producerImplementation.ProduceAsync(topicPartition, message, cancellationToken).ConfigureAwait(false);
```

Comment:
This activity should measure the duration of the `ProduceAsync` call. it should be disposed or stopped after `ProduceAsync` was awaited.

Reply:
done